### ISSUE TITLE
bundleconfig.json needs an update in order to enable bunding on save

### ISSUE THREAD

**Issue by MartinShishkov** [madskristensen/BundlerMinifier#351]
### Installed product versions
- Visual Studio 2017, version 15.6.1
- Bundler & Minifier, version 2.6.380

### Description
This is kinda strange behaviour and I don't really know how to file an issue for it but I'll try. 
First to say that I've already discovered a 'fix' but I don't know why it is working.

So, I had the bundle-on-save functionality not working for some reason. There weren't any changes to the project from the previous day (when it was working properly). I decided to update the extension but  it was still not compiling my bundles when I hit save. So the way I solved it was to make a minor change to bundleconfig.json file so I could SAVE it. After that the bundle-on-save started working again.

Any ideas on why is that? Is this even an issue with this extension or is it a VS bug in general?

**Comment by Luca7993** [comment]
I've had same problem.
I found that the extension has been disabled...after I reenable it, it still not working on save (tried also to close and reopen visual studio).
Make a minor change to file bundleconfig.json (just a space) solved it

**Comment by lepipele** [comment]
Looking at recent changes the only two commits who could potentially do something like this are:

https://github.com/madskristensen/BundlerMinifier/commit/bd6e202b42e0340605ea2397f59b9744d4fb90f6
https://github.com/madskristensen/BundlerMinifier/commit/d036a9168af809b950496c4d34ab0859a0ceeb3b

Can any of you clone repository and then do a bit of testing with compiling different DLLs (before these two commits) and then test? Like if it works before 1st or 2nd commit?

**Comment by MartinShishkov** [comment]
How are you reproducing this?

**Comment by daveylad** [comment]
I can confirm this is also an issue for me.. v2.6.378

changing the bundleconfig.json re-enables the update on save.  However, this needs to be done every time you restart VS

**Comment by spayton** [comment]
This started happening to me when I updated to vs2017 15.6.6.  I updated to the bundler nightly 2.6.380, that didn't help.  So uninstalled bundler and reinstalled it from vs to v2.6.378.

I also noticed intermittent VS crashes on file save, upon resume VS is telling me "A previous session terminated unexpectedly. Disabling extension Bundler & Minifier might help prevent similar issues...'

The suggestion above to edit the bundleconfig.json does get the bundling on save to start working and you do need to do this every VS launch.  Note that editing bundleconfig.json causes a bundle update when saved.

I'm still left with the intermittent crash on save though.

**Comment by lepipele** [comment]
@victorest03 and rest of you - I can't reproduce the issue... 

Just checked and my Visual Studio is 15.5.2. I also run it in admin mode, so that could make difference. What I suggest is that you do this:

1. Remove installed BundleMinifier extension, close Visual Studio
2. Clone this repository and set HEAD to commit 92df785749a995ce5a770a1b5e077967d1c9a242 as @victorest03 [mentions in his post](https://github.com/madskristensen/BundlerMinifier/issues/351#issuecomment-383618196)
3. Open BundlerMinifier.sln in Visual Studio... right click on `BundlerMinifierVisx` project in Solution Explorer and go `Debug -> Start new instance`
4. This will give you new instance of VS that you can use to debug error. Now, open the project where you were having problems. Use breakpoints back in the VS that has `BundlerMinifier.sln` open to figure out how to reproduce
5. Navigate commit by commit after 92df785749a995ce5a770a1b5e077967d1c9a242, hunt down error, and send PR
6. Be cherished by everyone in this thread for solving the issue

I'll try updating my Visual Studio as soon as I have and try different configuration settings to see if I can reproduce.

Also, I don't use extension, but rather directly reference DLL through NuGet in projects where BundlerMinifier is used. See if switching to that from extension helps.

**Comment by rockstardev** [comment]
@duganzhang that makes sense. You wanna submit PR?

Also @spayton @victorest03 @MartinShishkov @daveylad @Luca7993 - you guys all reported this problem. Once PR is ready will you be able to pull changes, build custom DLL and then test if that resolves the issue on your machines?

When we confirm patch addresses the problem, I'll talk with @madskristensen to push it to NuGet gallery.

**Comment by spayton** [comment]
Yes, I can try to test a fix.  I just tried to build using your guidance posted above, but got a load type/namespace missing/not-found errors, starting in AdornmentLayer.cs with:

    CS0234  C# The type or namespace name 'Text' does not exist in the namespace

I'm not familiar with cs projects so if someone can point me in the right direction to save me some research, then I'll be in a position to test quicker.

**Comment by JacobMWebb** [comment]
>  So the way I solved it was to make a minor change to bundleconfig.json file so I could SAVE it. After that the bundle-on-save started working again.

This provided a temporary fix for the issue for me as long as I have Visual Studio open, thank you.

**Comment by fordagame** [comment]
@spayton  It is a bit tricky, because maybe the version of Visual Studio with which tool was created is older. In order to load it I have install as nuget Microsoft.VisualStudio.Editor 15.6 . Than I removed: Microsoft.VisualStudio.Shell.Immutable 10.0 and 14.0 . Than I added two more packages: Microsoft.VisualStduio.Shell.Interop.15.3.DesignTime and Microsoft.VisualStduio.Shell.Interop.15.6.DesignTime .
This worked for me.

**Comment by spayton** [comment]
@fordagame well, I added some stuff and removed some stuff, got the error/warnings down to just 6 but in trying to resolve those last few, the count has not gone up again.  I think I've screwed it right up now.

are you able to post the extension here?  not sure what the limitations are with attachments

**Comment by fordagame** [comment]
It seems that I can. I can upload only zip file, because VISX is not allowed.
[BundlerMinifierVsix.zip](https://github.com/madskristensen/BundlerMinifier/files/1961683/BundlerMinifierVsix.zip)

**Comment by spayton** [comment]
@fordagame I deleted project and checked it out again, all fresh.  I made the correct reference mods this time and got it to compile ok.  So I then moved to the latest commit, built that and installed it... it now works, it bundles on save, whereas I was expecting it to not work.

So then I try your version and it also works fine.

So then I go back and install the official version 2.6.378, which is where I started, and it's not working.

So in summary, my version built with commit d036a91 on my vs2017 community, now works but it doesn't have the fix you posted above.  Your version now also works.  The official version still doesn't work.

I'm not sure how to explain why the version I built works, maybe I've installed something trying to get it to build which now makes it happy!

**Comment by rockstardev** [comment]
@spayton as soon as we merge this fix here - new nightly will automatically build and you guys who suffer from this problem will be able to update to it.

@fordagame - the problem is that now this is async it seems lots of threads trigger and go before everything is ready. During debugging I also now get `WebCompilerTaskRunner.cs` throwing exception here:

https://github.com/madskristensen/BundlerMinifier/blob/d036a9168af809b950496c4d34ab0859a0ceeb3b/src/BundlerMinifierVsix/TaskRunner/WebCompilerTaskRunner.cs#L39

I would much rather solve this in a way that makes extension wait until solution is ready and THEN does initialization. I think @madskristensen actually wanted to do that with `await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);` but it seems that method is not working as expected.

So yeah, until we solve this problem - workaround is to uninstall extension and instead use BundlerMinifier NuGet dll - it should have these issues.

**Comment by fordagame** [comment]
@rockstardev I found the same problem with WebCompilerTaskRunner.cs . I actually forgot to mention it, but I am not pretty sure that it will affect something or just crash the first time it happens. And I am almost sure that it is not related with this issue at least. I think that its crash is not affecting anything when EnableTaskRunnerExplorer is true. It anyway just return null.

If we wait everything to be loaded, this will mean that Solution will already be loaded. Actually most likely you are right, that some synchronization should be done most likely, but OnSolutionLoaded method will not be triggered anyway (or at least I think that this event should be passed already). Thats why my suggestion is just once we initialize ProjectEventCommand.cs , to call this method manually if solution is already loaded. Actually this is the idea of code... but if we already load the extension after the event, we just need to ensure that event is called once extension is loaded.

**Comment by spayton** [comment]
@fordagame forgot to mention that my build was release, not debug.  Here it is if anyone wants to try it:

[BundlerMinifierVsix-spayton1.zip](https://github.com/madskristensen/BundlerMinifier/files/1962251/BundlerMinifierVsix-spayton1.zip)

**Comment by rockstardev** [comment]
Call for testing: @spayton @victorest03 @MartinShishkov @daveylad @Luca7993 @fordagame   @JacobMWebb - can any of you help with testing this PR: 

https://github.com/madskristensen/BundlerMinifier/pull/357#issuecomment-385796095

**Comment by rockstardev** [comment]
@fordagame the root cause of the problem was actually exactly that line in `WebCompilerTaskRunner` I mentioned. It was race condition - that line would be called FIRST before BundlerMinifierPackage is initialized.

Since `Options` variable was `null`, call would fail and result on exception thrown and silently swallowed (because we are on background thread).

When user does that save manually, then this exact method `ParseConfig` would be triggered again... but this time `BundlerMinifierPackage.Options` is initialized... so from there it works, as @MartinShishkov  mentioned in workaround.

I've solved this by now `await`ing package initialization... and this way it works properly.

**Comment by rockstardev** [comment]
Hmph... maybe it has to do with different Visual Studio versions. So, you are saying that manual trigger of `OnSolutionOpen()` solves the issue on your machine?

I can add that to the set of changes on #355 and co-author you... you OK with that @fordagame?

**Comment by fordagame** [comment]
Yup. This is the problem.

Maybe if your VS is loading slowly or is different version it may happen that it will work. If your VS is loading slowly than it may happen that actually OnSolutionOpen method is executed. But review the code. See that exactly this method (which I pointed few times already) is responsible for FileChanged event. This is the only thing which matter for this issue, to call this method. Saving config file is calling the same thing as OnSolutionOppened. Everything else is not connected at all with FileChanged event.

I am using the latest version of VS. But based on this solution I guess it is done on 1 Major version bellow.

I have added PR here https://github.com/madskristensen/BundlerMinifier/pull/358
Either use it or if you want put it in your PR. It doesn't matter really.

If you want, suggest me some way to chat with you and we can discuss this issue, instead of writing this here again.

**Comment by rockstardev** [comment]
Yeah, something is definitely different across environments. For me `OnSolutionOpened` triggers event without any problems.

But let's keep it simple - @madskristensen is already checking for duplicate entries:

https://github.com/madskristensen/BundlerMinifier/blob/d036a9168af809b950496c4d34ab0859a0ceeb3b/src/BundlerMinifierVsix/Commands/ProjectEventCommand.cs#L73

So, this extra call that will happen on machines like mine won't cause any problems. I'll add another commit and co-author you. I'll leave like 12 hours for another tester from the mentioned list to give it a try. If we have no volunteers by then, I'll merge changes to produce new nightly build for everyone to enjoy (and then we can deal with bug requests if needed ;).

As for space to chat - I'll talk with @madskristensen to maybe start something up. He authored so many of these extensions and it would be good if there was some place people can meet and chat about non-formal stuff, instead of piling bunch of text here in issues ;)

Thanks for all the help @fordagame

**Comment by fordagame** [comment]
No Problem. I am happy that I can help you.

Just to mention that it will not trigger twice. We have two situation here: https://github.com/madskristensen/BundlerMinifier/pull/358/commits/b57f56f97f453abd2a0df12c926362cbe4e11649
1)We register event before Solution is open than this code will not trigger, because if is false, but event will be triggered once it happen. 
2) Or if we 1st open solution, than code will be true and it will be called, but obviously we missed the event. So don't worry in both situation we call the code once.

**Comment by rockstardev** [comment]
Yeah, tried debugging with your change set and it triggers twice for me in debug. By the end of the constructor in:

            if (dte.Solution.IsOpen)
            {
                OnSolutionOpened();
            }

And then some 250ms later, I get call from that subscription from `_events.Opened`.

I've moved that manual trigger up in constructor before event subscription to hopefully prevent some weird race conditions (because we execute constructor in sync). But we are good either way - I just hope we can finally put this issue behind us.

**Comment by fordagame** [comment]
I guess that this is the problem of debugging, because actually it may slow down the process while you keep breakpoint on one or other place in the code. 250 ms is too much for 3 lines of code :). In reallity I don't think this is possible.  To be honest being in the end of the constructor may cause it to execute two times, but in the begining of constructor may cause it to execute 0 if we are precise. (If we are false on Solution Opened while it pass through my if and than it is opened before we subscribe to event).  But how I said, this is happening only on debug, because you are staying on breakpoint and this is running on background, while you wait on Breakpoint actually state of Solution is changing and it enter twice.

**Comment by spayton** [comment]
@fordagame @rockstardev I can try a .vsix if you post one.  There's not much to be had from me building it, as my last build works!

**Comment by rockstardev** [comment]
@spayton I've merged PR and new nightly is built: http://vsixgallery.com/extension/a0ae318b-4f07-4f71-93cb-f21d3f03c6d3/ (look at Update Date)

Try it out and let me know how it works for you.

**Comment by daveylad** [comment]
@rockstardev I've just tried the nightly and it's working fine for me now

**Comment by rockstardev** [comment]
@daveylad has spoken! Issue closed ;)

**Comment by spayton** [comment]
@rockstardev I've tried the nightly on vs2015.  This used to have an intermittent crash on save but worked most of the time.  Previously it would crash VS, but now it seems able to dump some info to the output window.  so I'd say its about the same.  Here's the 1st chunk of output.

```
02/05/2018 22:45:14: Processing file 'W:\tsg\projects\games\crankitup\game\bundleconfig.json' results in exception:
System.IO.IOException: The process cannot access the file 'W:\tsg\projects\games\crankitup\game\src\js\bundles\bundle.ss.openbet.fiddler.js' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.StreamWriter.CreateFile(String path, Boolean append, Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path, Boolean append, Encoding encoding, Int32 bufferSize, Boolean checkHost)
   at System.IO.File.InternalWriteAllText(String path, String contents, Encoding encoding, Boolean checkHost)
   at System.IO.File.WriteAllText(String path, String contents, Encoding encoding)
   at BundlerMinifier.BundleFileProcessor.ProcessBundle(String baseFolder, Bundle bundle) in C:\projects\bundlerminifier\src\BundlerMinifier.Core\Bundle\BundleFileProcessor.cs:line 120
   at BundlerMinifier.BundleFileProcessor.SourceFileChanged(String bundleFile, String sourceFile) in C:\projects\bundlerminifier\src\BundlerMinifier.Core\Bundle\BundleFileProcessor.cs:line 68
   at BundlerMinifierVsix.BundleService.<>c__DisplayClass11_0.<SourceFileChanged>b__0(Object o) in C:\projects\bundlerminifier\src\BundlerMinifierVSIX\BundleService.cs:line 169

```

I'll try vs2017 tomorrow.

**Comment by fordagame** [comment]
@spayton  The fix as was stated before won't fix the exception, that file is in used. The fix, should fix the problem where VS crash when THIS exception occur. The question is does your VS crash now or just throw this exception? I would make assumption that if you can post us result from output window, than probably VS didn't crash :).

The problem was that exception handler, with new async package, crashes (cos of possiblities of Dead lock). It was fixed in order to not crash. 

Now a bit wild guess, but I am certain that this should be the problem. The problem which you are stating about file in use is caused, when you have a lot of javascript files which should be bundled and you save them at once. Than Extension take all of them and try to bundle, but they are working sometimes on the same bundle and etc, and that cause extension multiple time to try to access same bundle and of course it crash with FileAccess exception. Of course this should be fixed as well if possible, but issue is not about that.

**Comment by rockstardev** [comment]
@spayton yeah - it seems your MalwareBytes locks your file. What you are saying is more related to #355 than this issue.

Either way, try using [Sysinternals Process Explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) to debug which process locks your `JS` file... but otherwise this is outside of scope of what we can do here ;).

**Comment by spayton** [comment]
@fordagame  @rockstardev nightly now installed on vs2017, it is now bundling on save.  I'll have a full for or working with that so will report back if any problems occur.

**Comment by sharpmt** [comment]
I'm still having this issue in VS2017 Ver: 15.9.12.
